typed-racket-lib/typed-racket/typecheck/tc-expr-unit.rkt

```diff
@@ -273,8 +273,35 @@
        (ret (opt-unconvert (tc-expr/t #'fun)
                            (syntax->list #'(formals ...))))]
       ;; let
-      [(let-values ([(name ...) expr] ...) . body)
-       (tc/let-values #'((name ...) ...) #'(expr ...) #'body expected)]
+      [(let-values bindings . body)
+       (define bindings*
+         (syntax-parse #'body
+           #:literal-sets (kernel-literals)
+           ;; special case: let-values that originates from an application of a
+           ;; kw function. we may need to ignore contract-related arguments
+           [((kw-app1 (kw-app2 cpce s-kp fn kpe kws num) ; see tc-app/tc-app-keywords.rkt
+                      kw-list
+                      (kw-app3 list . kw-arg-list)
+                      . *pos-args))
+            #:declare cpce (id-from 'checked-procedure-check-and-extract 'racket/private/kw)
+            #:declare s-kp (id-from 'struct:keyword-procedure 'racket/private/kw)
+            #:declare kpe  (id-from 'keyword-procedure-extract 'racket/private/kw)
+            #:declare kw-app1  (id-from '#%app 'racket/private/kw)
+            #:declare kw-app2  (id-from '#%app 'racket/private/kw)
+            #:declare kw-app3  (id-from '#%app 'racket/private/kw)
+            #:when (contract-neg-party-property #'fn) ; contracted
+            ;; ignore the rhs which refers to a contract-lifted definition
+            (syntax-parse #'bindings
+              [(c1 [(contact-lhs) contract-rhs] cs ...)
+               ;; give up on optimizing the whole let, part of it is missing type info
+               ;; (not that this expansion could be optimized anyway)
+               (register-ignored! form)
```

Comment:
That only affects code inside calls to kw+contract functions, though.
Fixing it would require the optimizer to be able to handle partially typed expressions, which would be a big change.